### PR TITLE
Refactoring of QR: stabilized Gram-Schmidt for split=1 and TS-QR for split=0 

### PR DESCRIPTION
**Pull request by mrfh92**
### Changes in the implementation of QR: 
* `split=1`: column-block-wise stabilized/modified Gram Schmidt; thus any shape allowed. 
* `split=0`: TS-QR ("tall skinny QR"); thus input must be tall-skinny in the sense that each local chunk of data has not less rows than columns. Removing this restriction can be considered as future work. 
* API is now more NumPy/PyTorch-like:  inputs are the input matrix and a string `mode`; the latter can be `reduced` and `r`, but we currently do not support `complete` or `raw`. (`complete` is available both in PyTorch and NumPy and should be considered as future work.) The arguments `overwrite_a`, `calc_q`, and `tiles_per_proc` are no more used; instead I have introduced `procs_to_merge` which is a hyperparameter related to TS-QR. 
* skipping the QR-related unit test (i.e. those from `test_qr.py` and some from `test_svdtools.py`) on AMD-hardware has been removed since the new PyTorch/ROCm-images support `torch.linalg.qr`. 

### Possible future work:
* remove tall-skinny assumption for split=0 by combining TS-QR with Gram-Schmidt-like loop 
* The choice `mode=complete` is available both in PyTorch and NumPy and therefore might be included in future versions, too. 

## Due Diligence

- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [X] unit tests: all split configurations tested
    - [X] unit tests: multiple dtypes tested
    - [X] documentation updated where needed

### CODE REVIEW

**Comment by mrfh92** [comment]
Suggested handling of the options: 

* batches of matrices allowed 
* mode{‘reduced’, ‘complete’, ‘r’, ‘raw’}, optional with default  "reduced"

**Comment by mrfh92** [comment]
@mtar fyi: your update of the PyTorch/ROCm-image on the AMD-runner allows us now to test QR on AMD, too :+1:

**Comment by mrfh92** [comment]
@ClaudiaComito @mtar hopefully, this resolves the previous problems with QR 
@JuanPedroGHM cb/report is still "Pending" although the benchmarks have been run successfully in gitlab and also report success has been finished ... is there some problem with this?

**Comment by JuanPedroGHM** [comment]
> @ClaudiaComito @mtar hopefully, this resolves the previous problems with QR @JuanPedroGHM cb/report is still "Pending" although the benchmarks have been run successfully in gitlab and also report success has been finished ... is there some problem with this?

Fixed, the CB does not require the assignee tag to be set any more.

**Comment by mrfh92** [comment]
@JuanPedroGHM thx :+1:

File: heat/core/linalg/qr.py [hunk 1]
```diff
@@ -4,1039 +4,338 @@
...
-        if rank == pr0:
-            upper_rest = r_tiles.arr.lloc[upper_inds[0] : upper_inds[1], upper_inds[3] :]
-            lower_rest = torch.zeros(
-                lower_rest_size, dtype=r_tiles.arr.dtype.torch_type(), device=a_torch_device
+    if A.split == 0:
+        # implementation of TS-QR for split = 0
+        # check that data distribution is reasonable for TS-QR (i.e. tall-skinny matrix with also tall-skinny local chunks of data)
+        if A.lshape_map[:, 0].max().item() < A.shape[1]:
+            raise ValueError(
+                "A is split along the rows and the local chunks of data are rectangular with more rows than columns. \n Applying TS-QR in this situation is not reasonable w.r.t. runtime and memory consumption. \n We recomment to split A along the columns instead. \n In case this is not an option for you, please open an issue on GitHub."
```
Comment (ClaudiaComito) [inline heat/core/linalg/qr.py:171]: This is going to be weird for the users. You can start with a global TS matrix and end up with non-TS local arrays if you use enough processes. We need to explain it better / introduce an earlier check, maybe make a helper function available that calculates the most reasonable split for  given matrix size and nr of processes? I'll think about it some more.

File: heat/core/linalg/qr.py [hunk 2]
```diff
@@ -4,1039 +4,338 @@
...
-            sum_row = sum(q0_tiles.tile_rows_per_process[:pr])
-            end_row = q0_tiles.tile_rows_per_process[pr] + sum_row
-            # slice of q_tiles -> [0: -> end local, 1: start -> stop]
-            q_rest_loc = q0_tiles.local_get(key=(slice(None), slice(sum_row, end_row)))
-            # apply the local merge to q0 then update q0`
-            q_rest_loc = q_rest_loc @ local_merge_q[pr][0]
-            q0_tiles.local_set(key=(slice(None), slice(sum_row, end_row)), value=q_rest_loc)
-            del local_merge_q[pr]
+# -----------------------------------------------------------------------------
+# old version with serial sends instead of broadcast...
```
Comment (ClaudiaComito) [inline heat/core/linalg/qr.py:288]: dead code? (below too)

File: heat/core/linalg/qr.py [hunk 3]
```diff
@@ -4,1039 +4,338 @@
...
+                    Q_buf = Q_curr
+                if mode == "reduced":
+                    Q.larray = Q_curr
+                r_size = R.larray[R_shapes[i] : R_shapes[i + 1], :].shape[0]
+                R.larray[R_shapes[i] : R_shapes[i + 1], :] = R_loc[:r_size, :]
+
+            if i < nprocs - 1:
+                # broadcast the orthogonalized block of columns to all other processes
+                req = A.comm.Ibcast(Q_buf, root=i)
+                req.Wait()
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/linalg/qr.py:150]: Replacing these two lines by `A.comm.Bcast(Q_buf,root=i)` yields `MPI_ERR_TRUNCATE`

File: heat/core/linalg/qr.py [hunk 4]
```diff
@@ -4,1039 +4,338 @@
...
-    r_tiles = SquareDiagTiles(arr=r, tiles_per_proc=tiles_per_proc)
-    tile_columns = r_tiles.tile_columns
-    tile_rows = r_tiles.tile_rows
-    if calc_q:
-        q = factories.eye(
-            (r.gshape[0], r.gshape[0]), split=0, dtype=r.dtype, comm=r.comm, device=r.device
+    if not A.is_distributed():
+        # handle the case of a single process or split=None: just PyTorch QR
+        Q, R = torch.linalg.qr(A.larray, mode=mode)
+        R = DNDarray(
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/linalg/qr.py:76]: I think that for consistency it would be better to keep the same allowed modes for all split-variants

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Thanks for all the work, a few comments:

File: benchmarks/cb/linalg.py [hunk 5]
```diff
@@ -53,12 +51,16 @@ def run_linalg_benchmarks():
     matmul_split_1(a, b)
     del a, b
 
+    n = int((4000000 // MPI.COMM_WORLD.size) ** 0.5)
+    m = MPI.COMM_WORLD.size * n
+    a_0 = ht.random.random((m, n), split=0)
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply benchmarks/cb/linalg.py:56]: as far as I understand, the function to be monitored is `def qr_split_1(a)`, so the generation of test data is already outside the monitored area

File: heat/core/linalg/qr.py [hunk 6]
```diff
@@ -4,1039 +4,290 @@
...
-    overwrite_a: bool = False,
+    A: DNDarray,
+    mode: str = "reduced",
+    procs_to_merge: int = 2,
 ) -> Tuple[DNDarray, DNDarray]:
     r"""
     Calculates the QR decomposition of a 2D ``DNDarray``.
-    Factor the matrix ``a`` as *QR*, where ``Q`` is orthonormal and ``R`` is upper-triangular.
+    Factor the matrix ``A`` as *QR*, where ``Q`` is orthonormal and ``R`` is upper-triangular.
     If ``calc_q==True``, function returns ``QR(Q=Q, R=R)``, else function returns ``QR(Q=None, R=R)``
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/linalg/qr.py:27]: Taken into account. Added notice on how to replace calc_q under "notes".

File: heat/core/linalg/qr.py [hunk 7]
```diff
@@ -4,1039 +4,290 @@
...
+        default "reduced" returns Q and R with dimensions (M, min(M,N)) and (min(M,N), N), respectively.
+        "r" returns only R, with dimensions (min(M,N), N).
+    procs_to_merge : int, optional
+        determines the number of processes to be merged at one step during TS-QR (split = 0 only). Default is 2.
+        Higher choices may result in higher memory consumption. 0 corresponds to merging all processes at once.
 
     Notes
     -----
-    This function is built on top of PyTorch's QR function. ``torch.linalg.qr()`` using LAPACK on
+    For split=1, Q and R have split=1 as well. For split=0, Q has split=0 and R has split=None.
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/linalg/qr.py:43]: added

File: heat/core/linalg/qr.py [hunk 8]
```diff
@@ -4,1039 +4,290 @@
...
+    Other than ``numpy.linalg.qr()`` we only support ``mode="reduced"`` or ``mode="r"`` for the moment, since "complete" may result in heavy memory usage.
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/linalg/qr.py:44]: have taken the reformulation into account. 

communication overhead? -> thats not the real problem here. the problem is just that, e.g., you can have a 10^9 x 10 matrix easily in memory and you obtain a Q of shape 10^9 x 10 for mode="reduced", but a 10^9 x 10^9-shaped Q  for mode="complete". in short: for "complete", the size in memory of the QR can be an arbitrarily large multiple of the size of the original data, depending on the shape of the matrix

File: heat/core/linalg/qr.py [hunk 9]
```diff
@@ -4,1039 +4,290 @@
...
-    overwrite_a: bool = False,
+    A: DNDarray,
+    mode: str = "reduced",
+    procs_to_merge: int = 2,
 ) -> Tuple[DNDarray, DNDarray]:
     r"""
     Calculates the QR decomposition of a 2D ``DNDarray``.
-    Factor the matrix ``a`` as *QR*, where ``Q`` is orthonormal and ``R`` is upper-triangular.
+    Factor the matrix ``A`` as *QR*, where ``Q`` is orthonormal and ``R`` is upper-triangular.
     If ``calc_q==True``, function returns ``QR(Q=Q, R=R)``, else function returns ``QR(Q=None, R=R)``
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/linalg/qr.py:28]: I tried to clarify this.

File: heat/core/linalg/qr.py [hunk 10]
```diff
@@ -4,1039 +4,290 @@
...
-        If ``False``, a new array will be created for R
+    A : DNDarray of shape (M, N)
+        Array which will be decomposed. So far only 2D arrays with datatype float32 or float64 are supported
+        For split=0, the matrix must be tall skinny, i.e. the local chunks of data must have at least as many rows as columns.
+    mode : str, optional
+        default "reduced" returns Q and R with dimensions (M, min(M,N)) and (min(M,N), N), respectively.
+        "r" returns only R, with dimensions (min(M,N), N).
+    procs_to_merge : int, optional
+        determines the number of processes to be merged at one step during TS-QR (split = 0 only). Default is 2.
+        Higher choices may result in higher memory consumption. 0 corresponds to merging all processes at once.
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/linalg/qr.py:39]: I think its not possible to explain this in sufficient detail in the docs, but I added a warning that only users familiar with the algorithm should play around with this parameter.

**Comment by mrfh92** [comment]
I have addressed all except for the last comment; the last one I didnt understand because I dont see the inconsistency that is not there (?).

**Review by ClaudiaComito — APPROVED** [review]
Awesome @mrfh92 thank you so much!

**Pull request merged**